### ISSUE TITLE
Reduce boilerplate

### ISSUE THREAD

**Issue by JorgeBucaran** [hyperapp/hyperapp#489]
Instead of changing the API #484, let's approach and try to come up with a solution for the boilerplate problem within the current architecture.

## Importing/Exporting modules

```js
// ABC.js
import { B } from "./B"
import { C } from "./C"

export const state = {
  value: 0,
  B: B.state,
  C: C.state
}
export const actions = {
  add: value => state => ({ value: state.value + value }),
  B: B.actions,
  C: C.actions
}
```

## Setup

```js
import  { A } from "./A"
import  { B } from "./B"
import  { C } from "./C"

app({
  state: {
    A: A.state,
    B: B.state,
    C: C.state
  },
  actions: {
    A: A.actions,
    B: B.actions,
    C: C.actions
  }
})
```

## Props
```jsx
const view = state => actions => (
  <App>
    <Feed state={state.feed} actions={actions.feed} />
    <Trend state={state.trend} actions={actions.trend} />
  </App>
)

// OR

const view = state => actions => (
  <App>
    <Feed
      model={{
        ...state.feed,
        ...actions.feed
      }}
    />
    <Trend
      model={{
        ...state.trend,
        ...actions.trend
      }}
    />
  </App>
)
```
@lukejacksonn @SkaterDad @okwolf @zaceno @Alber70g

**Comment by JorgeBucaran** [comment]
@lukejacksonn @SkaterDad @okwolf @zaceno @Alber70g

**Comment by Mytrill** [comment]
A possible way to reduce boilerplate when composing modules/models, while still keeping state and action separate, could be to introduce an action that returns the initial state:
```javascript
// ABC.js
import { B } from "./B"
import { C } from "./C"

export const actions = {
  getInitialState: () => ({ value: 0 })
  add: value => state => ({ value: state.value + value }),
  B: B.actions,
  C: C.actions
}
```
```javascript
import  { A } from "./A"
import  { B } from "./B"
import  { C } from "./C"

app({
  actions: {
    A: A.actions,
    B: B.actions,
    C: C.actions
  }
})
```
The `getInitialState()` action would be executed automatically by `app()`, it could also be renamed `getState()` `init()` or simply `state()`.
This action could also be used to subscribe to outside events (like `init()`/`subscribe()` was meant to).

Not sure this is a good idea, let's see what you guys think.

**Comment by Alber70g** [comment]
@JorgeBucaran would it be possible to provide actions with the 'unmerged' state and the view with the model?

**Comment by okwolf** [comment]
@Alber70g the key question here is how strongly do we feel about keeping the API consistent and simple by minimizing the number of concepts new users need to learn.

**Comment by JorgeBucaran** [comment]
@whaaaley and I came up with an improvement to the current API without need to merge state & actions, but making it easier to reduce boilerplate.

It consists of extracting the view from the props argument to app.

```jsx
app(props, view, container)
```

Props is the state and actions, same API as 0.16.2. While we are at it, I find it more logical to call it `model`, instead of props, so I'll do just that.

```jsx
const model = {
  state: {
    count: 0
  },
  actions: {
    down: () => state => ({ count: model.count - 1 }),
    up: () => state => ({ count: model.count + 1 })
  }
}

const view = state => actions => (
  <main>
    <h1>{model.count}</h1>
    <button onclick={model.down}>-</button>
    <button onclick={model.up}>+</button>
  </main>
)

app(model, view)
```

So almost nothing changes, but the advantage of this approach is that it makes it now easier to merge multiple models.

```jsx
app(
  combine({ A, B, C, D }),
  view,
  document.getElementById("root")
)
```

or

```jsx
app(
  {
    ...model,
    ...combine({ A, B, C, D })
  },
  view,
  document.getElementById("root")
)
```

Of course you need to write that custom `combine` function, but surely this can be pushed to userland.

**Comment by JorgeBucaran** [comment]
- imperative

```jsx
const combine = models => {
  var out = {state:{},actions:{}}
  for (var key in models) {
    out.state[key] = models[key].state
    out.actions[key] = models[key].actions
  }
  return out
}
```

- functional
```jsx
const combine = models =>
  Object.keys(models).reduce(
    (obj, key) => ({
      ...obj,
      ...{
        state: {
          ...obj.state,
          ...{ [key]: models[key].state }
        },
        actions: {
          ...obj.actions,
          ...{ [key]: models[key].actions }
        }
      }
    }),
    {}
  )
```

**Comment by zaceno** [comment]
Looks good but it would also make sense to change the view-signature while were at it:

```js
app({state, actions}, ({state, actions}) => h(....))
```

Symmetrical to the conventional signature of components, and IMO nicer and less confusing to newcomers than `state => actions => h(...)`

**Comment by JorgeBucaran** [comment]
@zaceno I am okay with either one (state, actions) or ({ state, actions }).

**Comment by JorgeBucaran** [comment]
@Alber70g @lukejacksonn @okwolf @whaaaley What do you think? :)

**Comment by JorgeBucaran** [comment]
> Symmetrical to the conventional signature of components...

Good point though! 🤔😆

**Comment by SkaterDad** [comment]
This is definitely a nice incremental improvement! 

I'm voting for changing the view signature also.  Dont really care which of the 2 proposed solutions, they're both nicer.

**Comment by jamen** [comment]
fwiw, I been talking to @whaaaley and I like this approach.  For the view signature I would prefer `(state, actions)` but dont care too much.

**Comment by Alber70g** [comment]
I like this proposal even more (as opposed to the previous) because it keeps the state and actions separated while at the same time it allows actions and state on slices to be more explicit entangled. 

It also allows to have actions still return _only_ state instead of state and actions. 

The only thing I feel is kind of obsolete is the _initial state_. Right now we export the state as a part of the model that can be combined by the _combine_-function. 

This doesn't have to be this way if we, like Redux does, call all _init_- actions first. The init action can return it's default state. This also removes a part of the cognitive load of thinking about initial-state, actions and the 'current' state in actions, as a saparate thing. 

My proposal would be to drop the state drop the initial _app_-call as a whole and provide a lifecycle hook which calls each function named _init()_ in the actions object. Starting from root and traversing trough the slices/props.

**Comment by JorgeBucaran** [comment]
@Alber70g I don't follow, what's obsolete?

> My proposal would be to drop the state drop the initial app-call as a whole and provide a lifecycle hook which calls each function named init() in the actions object. Starting from root and traversing trough the slices/props.

What would be the benefit of adding something like this potentially adding a lot more bytes to hyperapp?

**Comment by SkaterDad** [comment]
I'm not really a fan of that idea. Being able to define the initial state before Hyperapp starts is really nice, in my opinion, and doesn't introduce special framework specific actions.

**Comment by JorgeBucaran** [comment]
I think this is Hyperapp's final API.

```jsx
const model = {
  state: {
    count: 0
  },
  actions: {
    down: () => state => ({ count: state.count - 1 }),
    up: () => state => ({ count: state.count + 1 }),
    /* actionName: data => (state, actions) => State */
  }
}

const view = ({ state, actions }) => (
  <main>
    <h1>{state.count}</h1>
    <button onclick={actions.down}>-</button>
    <button onclick={actions.up}>+</button>
  </main>
)

export const App = app(model, view) // => { state, actions }
```

**Comment by SkaterDad** [comment]
Looks good.  Is the root element a 3rd parameter missing from the example?

**Comment by JorgeBucaran** [comment]
@SkaterDad Done.

Hmm, should we make it non optional? 🤔🤯

**Comment by SkaterDad** [comment]
I think it's nice that its optional currently, but I'm not that opinionated on it.

**Comment by JorgeBucaran** [comment]
@SkaterDad I feel the same way, but maybe this can somehow be connected with #491.

**Comment by okwolf** [comment]
@JorgeBucaran well, now that you have a feeling we should just ship 1.0 immediately 😂

**Comment by Swizz** [comment]
The optionated root element make some troubles in some case. 

This is not a secret now, I use Hyperapp to build web component, using the custom
element compliance `<hyperapp-custom-element/>`.

Using the new API, it will be something like : 
```jsx
document.querySelectorAll('hyperapp-custom-element).map(element => {
  app(model, view, element)
})
```
In the case of there is no `<hyperapp-custom-element/>`, nothing will occur (the array is empty, no map iteration is called).

But if I am only doing the following : 
```jsx
app(model, view, document.querySelector('#app'))
```
And no `#app` exist in the dom, the whole body will be replaced by the app.

#### tl;dr
I use a custom root element, because the app is located somewhere in a landing page, if I remove the element from the html body, I want the app do not load not to replace the whole body.

**Comment by Zhami** [comment]
I prefer the proposed view signature of `({state, actions})`. The current (0.16.2) approach of double fat arrows `state => actions => h()` to me implies/suggests that there is an opportunity for currying, but that is not really the case.

**Comment by whaaaley** [comment]
@Alber70g I like your way of thinking! We're on the same page as far as "only actions should make state". The caveat here is that you need a way to load in old state. Having a top-level state prop in `app()` lets us load in initial state from local storage or some other place.

**Comment by JorgeBucaran** [comment]
I am very happy with the result and feel this can be closed.

**Comment by Alber70g** [comment]
@whaaaley you can still do:

```js
const appActions app(state, {init: payload => ({ ...payload }) }, view);
appActions.init(storage.getItem('myState'));
```

If I'm not mistaken and the render function has a `setTimeout()`/is put on the callback queue, this wont even pose a render performance hit. Therefore you don't need to set the initial `state` as property of `app`.

**Comment by Alber70g** [comment]
@JorgeBucaran

It might be good to add a code example where different slices are exported as models and combined by the combine function?

**Comment by lukejacksonn** [comment]
My favourite bit about the models proposal was the way it made passing stuff down the view much easier.. and that when you called stuff it was by `namespace.value` not `(state|actions).namespace.value`. But yes.. you could write a custom combine function that turns almost any data structure into what hyperapp accepts.

**Comment by JorgeBucaran** [comment]
> @lukejacksonn My favourite bit about the models proposal was the way it made passing stuff down the view much easier.. and that when you called stuff it was by namespace.value not (state|actions).namespace.value.

My favorite bit as well. Almost like a `class`.
